aiida_workgraph/engine/workgraph.py

```diff
@@ -646,9 +682,27 @@ def update_task_state(self, name: str) -> None:
             self.task_set_context(name)
             self.report(f"Task: {name} finished.")
         else:
-            task["results"] = None
+            task.setdefault("results", None)
+
+        self.update_parent_task_state(name)
 
+    def set_normal_task_results(self, name, results):
+        """Set the results of a normal task."""
+        task = self.ctx._tasks[name]
+        if isinstance(results, tuple):
+            if len(task["outputs"]) != len(results):
+                return self.exit_codes.OUTPUS_NOT_MATCH_RESULTS
+            for i in range(len(task["outputs"])):
+                task["results"][task["outputs"][i]["name"]] = results[i]
+        elif isinstance(results, dict):
+            task["results"] = results
+        else:
+            task["results"][task["outputs"][0]["name"]] = results
+        self.set_task_state_info(name, "state", "FINISHED")
+        self.task_set_context(name)
+        self.report(f"Task: {name} finished.")
```

Comment:
I know a bit nitpicky, but shouldn't the order be?
```suggestion
        self.task_set_context(name)
        self.set_task_state_info(name, "state", "FINISHED")
        self.report(f"Task: {name} finished.")
```

I don't know if the task should be marked as finished if task_set_context fails, but this is the pattern I see here

Reply:
Thanks! Done!